File: src/boot-scene.js
class BootScene extends Phaser["Scene"] {
  constructor() {
    super({
      key: "BootScene",
    });
  }
  ["preload"]() {
    !(function (value72) {
      if (value72["renderer"]["type"] === Phaser["WEBGL"]) {
        let value73 = value72["renderer"]["gl"];
        ((blendAdd = value72["renderer"]["addBlendMode"](
          [value73["SRC_ALPHA"], value73["ONE"]],
          value73["FUNC_ADD"],
        )),
          (blendNormal = value72["renderer"]["addBlendMode"](
            [value73["DST_COLOR"], value73["ONE_MINUS_SRC_ALPHA"]],
            value73["FUNC_ADD"],
          )));
      }
    })(this["game"]);
    let width = this["cameras"]["main"]["width"],
      height = this["cameras"]["main"]["height"],
      value71 = 0.6 * width,
      rectangle = this["add"]
        ["rectangle"](width / 2, height / 2, value71, 8, 65280)
        ["setOrigin"](0.5, 0.5);
    ((rectangle["scaleX"] = 0),
      this["load"]["on"]("progress", (progress) => {
        rectangle["scaleX"] = progress;
      }),
      this["load"]["on"]("loaderror", (loaderror) => {}),
      this["load"]["atlas"](
        "GJ_WebSheet",
        "assets/GJ_WebSheet.png",
        "assets/GJ_WebSheet.json",
      ),
      this["load"]["image"]("bigFont", "assets/bigFont.png"),
      this["load"]["text"]("bigFontFnt", "assets/bigFont.fnt"),
      this["load"]["image"]("goldFont", "assets/goldFont.png"),
      this["load"]["text"]("goldFontFnt", "assets/goldFont.fnt"),
      this["load"]["image"]("game_bg_01", "assets/game_bg_01_001.png"),
      this["load"]["image"]("sliderBar", "assets/sliderBar.png"),
      this["load"]["image"]("square04_001", "assets/square04_001.png"),
      this["load"]["image"]("GJ_square02", "assets/GJ_square02.png"),
      this["load"]["text"]("level_1", "assets/1.txt"),
      this["load"]["audio"]("stereo_madness", "assets/StereoMadness.mp3"),
      this["load"]["audio"]("explode_11", "assets/explode_11.ogg"),
      this["load"]["audio"]("endStart_02", "assets/endStart_02.ogg"),
      this["load"]["audio"]("playSound_01", "assets/playSound_01.ogg"),
      this["load"]["audio"]("quitSound_01", "assets/quitSound_01.ogg"),
      this["load"]["audio"]("highscoreGet02", "assets/highscoreGet02.ogg"));
  }
  ["create"]() {
    this["cache"]["text"]["get"]("level_1");
    const text = this["cache"]["text"]["get"]("bigFontFnt");
    text && defineFontFromFnt(this, "bigFont", text);
    const text2 = this["cache"]["text"]["get"]("goldFontFnt");
    (text2 && defineFontFromFnt(this, "goldFont", text2),
      this["scene"]["start"]("GameScene"));
  }
}

File: src/boot-scene.cpp
#include "boot-scene.h"
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>

#define STB_IMAGE_IMPLEMENTATION
#include "stb_image.h"
#include "level-data-helpers.h"

// Forward declaration from font-helpers
void defineFontFromFnt(const std::string& fontKey, const std::string& fntText);

std::unordered_map<std::string, Texture> BootScene::textures;
std::unordered_map<std::string, std::string> BootScene::textCache;

BootScene::BootScene() {}
BootScene::~BootScene() {}

std::string BootScene::loadTextFile(const std::string& path) {
    std::ifstream file(path);
    if (!file.is_open()) {
        std::cerr << "[BootScene] Failed to open file: " << path << std::endl;
        return "";
    }
    std::stringstream buffer;
    buffer << file.rdbuf();
    return buffer.str();
}

Texture BootScene::loadTextureGL(const std::string& path) {
    Texture tex;
    int channels;
    stbi_set_flip_vertically_on_load(false);
    unsigned char* data = stbi_load(path.c_str(), &tex.width, &tex.height, &channels, 4);
    if (!data) {
        std::cerr << "[BootScene] Failed to load texture: " << path << std::endl;
        return tex;
    }

    glGenTextures(1, &tex.id);
    glBindTexture(GL_TEXTURE_2D, tex.id);

    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, GL_CLAMP);
    glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, GL_CLAMP);

    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA, tex.width, tex.height, 0, GL_RGBA, GL_UNSIGNED_BYTE, data);
    stbi_image_free(data);
    return tex;
}

void BootScene::renderProgressBar(float progress) {
    glClear(GL_COLOR_BUFFER_BIT);

    float barTotalWidth = 0.6f * screenWidth;
    float barHeight = 8.0f;
    float currentWidth = barTotalWidth * progress;

    float centerX = screenWidth * 0.5f;
    float centerY = screenHeight * 0.5f;

    float startX = centerX - (barTotalWidth * 0.5f);
    float startY = centerY - (barHeight * 0.5f);

    glDisable(GL_TEXTURE_2D);
    glColor3f(0.0f, 1.0f, 0.0f);

    glBegin(GL_QUADS);
    glVertex2f(startX, startY);
    glVertex2f(startX + currentWidth, startY);
    glVertex2f(startX + currentWidth, startY + barHeight);
    glVertex2f(startX, startY + barHeight);
    glEnd();

    glEnable(GL_TEXTURE_2D);
}

void BootScene::preload(SDL_Window* window) {
    struct AssetTask {
        std::string key;
        std::string path;
        bool isTexture;
    };

    std::vector<AssetTask> tasks = {
        {"GJ_WebSheet",    "assets/GJ_WebSheet.png",     true},
        {"bigFont",        "assets/bigFont.png",         true},
        {"bigFontFnt",     "assets/bigFont.fnt",         false},
        {"goldFont",       "assets/goldFont.png",        true},
        {"goldFontFnt",    "assets/goldFont.fnt",        false},
        {"game_bg_01",     "assets/game_bg_01_001.png",  true},
        {"sliderBar",      "assets/sliderBar.png",       true},
        {"square04_001",   "assets/square04_001.png",    true},
        {"GJ_square02",    "assets/GJ_square02.png",     true},
        {"GJ_WebSheetJson","assets/GJ_WebSheet.json",    false},
        {"level_1",        "assets/1.txt",               false}
    };

    for (size_t i = 0; i < tasks.size(); ++i) {
        if (tasks[i].isTexture) {
            textures[tasks[i].key] = loadTextureGL(tasks[i].path);
        } else {
            textCache[tasks[i].key] = loadTextFile(tasks[i].path);
        }

        float progress = (float)(i + 1) / (float)tasks.size();
        renderProgressBar(progress);
        SDL_GL_SwapWindow(window);
    }
}

void BootScene::create() {
    if (textCache.find("bigFontFnt") != textCache.end()) {
        defineFontFromFnt("bigFont", textCache["bigFontFnt"]);
    }
    if (textCache.find("goldFontFnt") != textCache.end()) {
        defineFontFromFnt("goldFont", textCache["goldFontFnt"]);
    }

    std::cout << "[BootScene] Preload completed successfully." << std::endl;
}
